### PR TITLE
ZMQ based state tables

### PR DESCRIPTION
**Pull request by liuh-80**
#### Why I did it
Add ZMQ based ProducerStateTable and CustomerStateTable, which is 15 times faster than Redis based tables.
For transfer 10000 route config, ZMQ based tables take 120ms, Redis based tables take 1796ms. 

#### How I did it
Add ZMQ based ProducerStateTable and CustomerStateTable.

#### How to verify it
Add new UT to cover ZmqProducerStateTable and ZmqCustomerStateTable.
Pass all UT and E2E test cases.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

#### Description for the changelog
Add ZMQ based ProducerStateTable and CustomerStateTable.

### CODE REVIEW

File: common/zmqproducerstatetable.cpp [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+    for (const auto &key : keys)
+    {
+        sendMsg(key, vector<FieldValueTuple>(), DEL_COMMAND);
+    }
+}
+
+void ZmqProducerStateTable::updateTableThreadFunction()
+{
+    DBConnector db(m_pipe->getDbName(), 0, true);
```
Comment (liuh-80) [inline common/zmqproducerstatetable.cpp:160]: WHen check some UT failed issue, I found write temp record to Redis seems not necessary, will update this PR and remove these code.

Comment (liuh-80) [reply common/zmqproducerstatetable.cpp]: Fixed

File: common/zmqconsumerstatetable.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        SWSS_LOG_THROW("zmq_bind failed on endpoint: %s, zmqerrno: %d",
+                m_endpoint.c_str(),
+                zmq_errno());
+    }
+
+    SWSS_LOG_NOTICE("bind to zmq endpoint: %s", m_endpoint.c_str());
+    while (m_runThread)
+    {
+        // receive message
+        rc = zmq_recv(socket, buffer.data(), MQ_RESPONSE_BUFFER_SIZE, ZMQ_DONTWAIT);
```
Comment (liuh-80) [inline common/zmqconsumerstatetable.cpp:99]: need improve performance by use zmq_poll

Comment (liuh-80) [reply common/zmqconsumerstatetable.cpp:115]: Fixed

**Comment by liuh-80** [comment]
Code change been validated with this PR:
https://github.com/sonic-net/sonic-buildimage/pull/12637

File: common/producerstatetable.h [hunk 3]
```diff
@@ -12,7 +12,7 @@ class ProducerStateTable : public TableBase, public TableName_KeySet
 public:
     ProducerStateTable(DBConnector *db, const std::string &tableName);
     ProducerStateTable(RedisPipeline *pipeline, const std::string &tableName, bool buffered = false);
-    ~ProducerStateTable();
+    virtual ~ProducerStateTable();
```
Comment (qiluo-msft) [inline common/producerstatetable.h:15]: >virtual [](http://example.com/codeflow?start=4&length=7)

This is valid fix, and could be separated into another PR.

File: common/producerstatetable.h [hunk 4]
```diff
@@ -50,7 +50,7 @@ class ProducerStateTable : public TableBase, public TableName_KeySet
     void create_temp_view();
 
     void apply_temp_view();
-private:
+protected:
```
Comment (qiluo-msft) [inline common/producerstatetable.h:53]: >protected [](http://example.com/codeflow?start=0&length=9)

Do you reuse anything of ProducerStateTable in ZmqProducerStateTable? If not, do not inherit it.

File: common/zmqproducerstatetable.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+#include <queue>
+#include <thread> 
+#include <mutex> 
+#include "table.h"
+#include "redispipeline.h"
+#include "producerstatetable.h"
+
+namespace swss {
+
+class ZmqProducerStateTable : public ProducerStateTable
```
Comment (qiluo-msft) [inline common/zmqproducerstatetable.h:14]: >ProducerStateTable [](http://example.com/codeflow?start=37&length=18)

The 'State' in 'ProducerStateTable' means that set/del operation on the same key could be combined if they are accumulated in the channel and the consumer is slow.

Do you have similar behavior? if not, just call it ZmqProducerTable

File: common/producerstatetable.h (hunk 3 above)
Comment (liuh-80) [reply common/producerstatetable.h:15]: Fixed, create new PR for this, will update after that PR merged.

https://github.com/sonic-net/sonic-swss-common/pull/735/files

File: common/zmqproducerstatetable.h (hunk 5 above)
Comment (liuh-80) [reply common/zmqproducerstatetable.h:14]: Yes, these new classes following same logic: after consumer table receive all data, the set/del operation on same key will combined.

File: common/producerstatetable.h (hunk 4 above)
Comment (liuh-80) [reply common/producerstatetable.h]: Fixed.

File: common/zmqproducerstatetable.cpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        }
+
+        int rc = zmq_send(m_socket, m_sendbuffer.data(), serializedlen, ZMQ_DONTWAIT);
+        if (rc >= 0)
+        {
+            SWSS_LOG_DEBUG("zmq sended %d bytes", serializedlen);
+            return;
+        }
+
+        if (zmq_errno() == EINTR
```
Comment (qiluo-msft) [inline common/zmqproducerstatetable.cpp:160]: >zmq_errno [](http://example.com/codeflow?start=12&length=9)

Store the return errno in a local var.

File: common/zmqproducerstatetable.cpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    }
+}
+    
+void ZmqProducerStateTable::initialize(const std::string& endpoint)
+{
+    m_connected = false;
+    m_endpoint = endpoint;
+    m_context = nullptr;
+    m_socket = nullptr;
+    m_sendbuffer.resize(MQ_RESPONSE_BUFFER_SIZE);
```
Comment (qiluo-msft) [inline common/zmqproducerstatetable.cpp:56]: >MQ_RESPONSE_BUFFER_SIZE [](http://example.com/codeflow?start=24&length=23)

This is not "sizes" but kind of maximal count.

File: common/binaryserializer.h [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        return finalize();
+    }
+
+    static void deSerializeBuffer(const char* buffer, const size_t size, std::vector<swss::FieldValueTuple>& values)
+    {
+        WARNINGS_NO_CAST_ALIGN
+        auto pkvp_count = (const size_t*)buffer;
+        WARNINGS_RESET
```
Comment (qiluo-msft) [inline common/binaryserializer.h:38]: >WARNINGS_RESET [](http://example.com/codeflow?start=8&length=14)

Add `;` at end?

File: common/binaryserializer.h [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        // return size
+        return m_current_position - m_buffer;
+    }
+
+    void setData(const char* data, size_t datalen)
+    {
+        if ((size_t)(m_current_position - m_buffer + datalen + sizeof(size_t)) > m_buffer_size)
+        {
+            SWSS_LOG_THROW("There are not enough buffer for binary serializer to serialize, key count: %zu, data length %zu, buffer size: %zu",
+                                                                                                                                                m_kvp_count,
```
Comment (qiluo-msft) [inline common/binaryserializer.h:120]: >                                                                                                                                                 [](http://example.com/codeflow?start=0&length=144)

too many indentation

File: common/binaryserializer.h [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            auto& field = fvField(kvp);
+            auto& value = fvValue(kvp);
+            setKeyAndValue(field.c_str(), field.length(), value.c_str(), value.length());
+        }
+
+        return finalize();
+    }
+
+    static void deSerializeBuffer(const char* buffer, const size_t size, std::vector<swss::FieldValueTuple>& values)
```
Comment (qiluo-msft) [inline common/binaryserializer.h:34]: >deSerializeBuffer [](http://example.com/codeflow?start=16&length=17)

deSerializeBuffer -> deserializeBuffer

Comment (qiluo-msft) [inline common/binaryserializer.h:34]: >static [](http://example.com/codeflow?start=4&length=6)

Why this function is `static` but above one is not?

File: common/zmqproducerstatetable.cpp [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+            SWSS_LOG_DEBUG("zmq sended %d bytes", serializedlen);
+            return;
+        }
+
+        if (zmq_errno() == EINTR
+            || zmq_errno() == EAGAIN
+            || zmq_errno() == EFSM)
+        {
+            // EINTR: interrupted by signal
+            // EAGAIN: ZMQ is full to need try again
```
Comment (qiluo-msft) [inline common/zmqproducerstatetable.cpp:165]: >ZMQ is full to need try again [](http://example.com/codeflow?start=23&length=29)

This is a serious problem. Maybe downstream component hangs or crashes. We need have at least warning logging.

File: common/zmqproducerstatetable.cpp [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+            // EFSM: socket state not ready
```
Comment (qiluo-msft) [inline common/zmqproducerstatetable.cpp:166]: >socket state not ready [](http://example.com/codeflow?start=21&length=22)

When will this happen? Only at initialization?

File: common/zmqproducerstatetable.cpp [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+            // for more detail, please check: http://api.zeromq.org/2-1:zmq-send
+            SWSS_LOG_DEBUG("zmq send retry, endpoint: %s, error: %d", m_endpoint.c_str(), zmq_errno());
+        }
+        else if (zmq_errno() == ETERM)
```
Comment (qiluo-msft) [inline common/zmqproducerstatetable.cpp:170]: >ETERM [](http://example.com/codeflow?start=32&length=5)

When will it happen? Is it better to throw, which is same as ProducerStateTable?

File: common/zmqproducerstatetable.cpp [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+            SWSS_LOG_DEBUG("zmq connection break, endpoint: %s,error: %d", m_endpoint.c_str(), rc);
+        }
+        else
+        {
+            // for other error, send failed immediately.
+            SWSS_LOG_THROW("zmq send failed, endpoint: %s, error: %d", m_endpoint.c_str(), rc);
+        }
+
+        // sleep (retry time) * 10 ms
+        usleep((i+1) * 10 * 1000);
```
Comment (qiluo-msft) [inline common/zmqproducerstatetable.cpp:183]: >usleep [](http://example.com/codeflow?start=8&length=6)

Prefer double the backoff time each iteration. Need to be more verbose in syslog in a nearly infinite loop.

File: common/zmqconsumerstatetable.h [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+ 
+    static inline std::shared_ptr<KeyOpFieldsValuesTuple> deserializeReceivedData(const char* buffer, const size_t size);
+
+    void mqPollThread();
+
+    void dbUpdateThread();
+
+    volatile bool m_runThread;
+
+    std::shared_ptr<std::thread> m_mqPollThread;
```
Comment (qiluo-msft) [inline common/zmqconsumerstatetable.h:95]: >thread [](http://example.com/codeflow?start=25&length=6)

We need to have a design doc to understand:
1. how do the threads behave?
2. what are the failure models with ZMQ as a outside dependency, and how should the class handle the failures.

File: common/zmqproducerstatetable.cpp (hunk 12 above)
Comment (liuh-80) [reply common/zmqproducerstatetable.cpp:166]: For example when ZMQ socket still not receive reply message from last sended package.
There was state machine inside ZMQ socket, when the socket is not in ready to send state, this error will happen.
Will add this information to code comments.

File: common/zmqproducerstatetable.cpp (hunk 13 above)
Comment (liuh-80) [reply common/zmqproducerstatetable.cpp:170]: ZmqProducerStateTable is a client, when server side down, this error will happen.
For example when orchagent restart, gnmi side ZmqProducerStateTable will receive this error, then gnmi side need reconnect and send again. if failed after retry ZmqProducerStateTable will throw.

File: common/zmqconsumerstatetable.h (hunk 15 above)
Comment (liuh-80) [reply common/zmqconsumerstatetable.h:95]: Will create a design doc ins SONiC repo later.

File: common/zmqproducerstatetable.cpp (hunk 6 above)
Comment (liuh-80) [reply common/zmqproducerstatetable.cpp:160]: Fixed

File: common/zmqproducerstatetable.cpp (hunk 7 above)
Comment (liuh-80) [reply common/zmqproducerstatetable.cpp:56]: Fixed, rename to MQ_RESPONSE_MAX_COUNT

File: common/binaryserializer.h (hunk 8 above)
Comment (liuh-80) [reply common/binaryserializer.h:38]: Fixed

File: common/binaryserializer.h (hunk 9 above)
Comment (liuh-80) [reply common/binaryserializer.h:120]: Fixed

File: common/binaryserializer.h (hunk 10 above)
Comment (liuh-80) [reply common/binaryserializer.h:34]: Fixed

Comment (liuh-80) [reply common/binaryserializer.h:34]: Fixed.
Serialize method difficult to change to fully static code, because need share current buffer position across multiple methods.

File: common/zmqproducerstatetable.cpp (hunk 11 above)
Comment (liuh-80) [reply common/zmqproducerstatetable.cpp:165]: Fixed, added warning log.

File: common/zmqproducerstatetable.cpp (hunk 14 above)
Comment (liuh-80) [reply common/zmqproducerstatetable.cpp:183]: Fixed.
Will retry and throw after retry failed.

**Comment by prvattem** [comment]
Is there a design proposal for these changes? If yes could you please share it across?

**Comment by liuh-80** [comment]
> Is there a design proposal for these changes? If yes could you please share it across?

Here is design document for this PR:

https://github.com/sonic-net/SONiC/pull/1215

**Review by marian-pritsak — DISMISSED** [review]

File: common/zmqproducerstatetable.cpp (hunk 14 above)
Comment (qiluo-msft) [reply common/zmqproducerstatetable.cpp]: `pow` is too heavy. Could you just x2 each iteration?

File: common/zmqproducerstatetable.cpp [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+        int rc = zmq_send(m_socket, m_sendbuffer.data(), serializedlen, ZMQ_DONTWAIT);
+        if (rc >= 0)
+        {
+            SWSS_LOG_DEBUG("zmq sended %d bytes", serializedlen);
+            return;
+        }
+
+        zmq_err = zmq_errno();
+        // sleep (z ^ retry time) * 10 ms
```
Comment (qiluo-msft) [inline common/zmqproducerstatetable.cpp:167]: >z [](http://example.com/codeflow?start=18&length=1)

z -> 2

File: common/zmqproducerstatetable.cpp (hunk 13 above)
Comment (qiluo-msft) [reply common/zmqproducerstatetable.cpp]: The "example when orchagent restart" is interesting. In this case, retry at low level may be not the best choice at high level. It is better to throw or return error to caller, and let caller make high level decision. For example: tell the gNMI client that sonic restarted, and further decision like retry one entry or retry whole table should be made at client level.

File: common/zmqproducerstatetable.cpp (hunk 16 above)
Comment (liuh-80) [reply common/zmqproducerstatetable.cpp]: Fixed

File: common/zmqproducerstatetable.cpp (hunk 13 above)
Comment (qiluo-msft) [reply common/zmqproducerstatetable.cpp]: I do not see how "if failed after retry ZmqProducerStateTable will throw."

Comment (liuh-80) [reply common/zmqproducerstatetable.cpp]: Fixed, the 'SWSS_LOG_THROW' will thorw swss exception.
Code change to throw exception when connection break, and caller need check connection status and connect again.

File: common/zmqproducerstatetable.cpp (hunk 14 above)
Comment (liuh-80) [reply common/zmqproducerstatetable.cpp]: Fixed.

Comment (qiluo-msft) [reply common/zmqproducerstatetable.cpp]: Sorry for misleading. I mean you can achieve `pow` by `retry_delay *= 2`. The only improvement is on performance, and the behavior is still double backoff.

Comment (liuh-80) [reply common/zmqproducerstatetable.cpp]: Fixed, this method moved to ZmqClient

File: common/zmqclient.cpp [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+            retry_delay = 0;
+        }
+        else if (zmq_err == EAGAIN)
+        {
+            // EAGAIN: ZMQ is full to need try again
+            SWSS_LOG_WARN("zmq is full, will retry in %d ms, endpoint: %s, error: %d", retry_delay, m_endpoint.c_str(), zmq_err);
+        }
+        else if (zmq_err == ETERM)
+        {
+            // reconnect and send again.
```
Comment (qiluo-msft) [inline common/zmqclient.cpp:156]: >reconnect and send again [](http://example.com/codeflow?start=15&length=24)

If it throw, the comment is not up-to-date "reconnect and send again".

File: common/zmqconsumerstatetable.cpp [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+    return clone;
+}
+
+void ZmqConsumerStateTable::handleReceivedData(std::shared_ptr<KeyOpFieldsValuesTuple> pkco)
+{
+    std::shared_ptr<KeyOpFieldsValuesTuple> clone = nullptr;
+    if (m_dbUpdateThread != nullptr)
+    {
+        // clone before put to received queue, because received data may change by consumer.
+        clone = cloneKeyOpFieldsValuesTuple(pkco);
```
Comment (qiluo-msft) [inline common/zmqconsumerstatetable.cpp:80]: >cloneKeyOpFieldsValuesTuple [](http://example.com/codeflow?start=16&length=27)

Is the default constructor of KeyOpFieldsValuesTuple good enough? why need a clone function?

File: common/zmqconsumerstatetable.h [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+
+namespace swss {
+
+class ZmqConsumerStateTable : public Selectable, ZmqMessageHandler
+{
+public:
+    /* The default value of pop batch size is 128 */
+    static constexpr int DEFAULT_POP_BATCH_SIZE = 128;
+
+    ZmqConsumerStateTable(DBConnector *db, const std::string &tableName, ZmqServer& zmqServer, int popBatchSize = DEFAULT_POP_BATCH_SIZE, int pri = 0, bool dbPersistence = true);
```
Comment (qiluo-msft) [inline common/zmqconsumerstatetable.h:25]: >& [](http://example.com/codeflow?start=82&length=1)

Minor format issue: the placement of `&` is not consistent.

File: common/zmqserver.cpp [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+    m_runThread = false;
+    m_mqPollThread->join();
+}
+
+void ZmqServer::registerMessageHandler(
+                                    const std::string dbName,
+                                    const std::string tableName,
+                                    ZmqMessageHandler* handler)
+{
+    auto dbMappingIter = m_HandlerMap.find(dbName);
```
Comment (qiluo-msft) [inline common/zmqserver.cpp:34]: >find [](http://example.com/codeflow?start=38&length=4)

you can call `insert(dbName)` and return the pair of dbMappingIter and bool.

File: common/zmqserver.cpp [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+
+void ZmqServer::mqPollThread()
+{
+    SWSS_LOG_ENTER();
+    SWSS_LOG_NOTICE("mqPollThread begin");
+    std::vector<char> buffer;
+    buffer.resize(MQ_RESPONSE_MAX_COUNT);
+
+    // Producer/Consumer state table are n:1 mapping, so need use PUSH/PULL pattern http://api.zeromq.org/master:zmq-socket
+    void* context = zmq_ctx_new();;
```
Comment (qiluo-msft) [inline common/zmqserver.cpp:104]: >zmq_ctx_new [](http://example.com/codeflow?start=20&length=11)

Do you also need mutex?

File: common/zmqserver.cpp [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+    int rc = zmq_bind(socket, m_endpoint.c_str());
+    if (rc != 0)
+    {
+        SWSS_LOG_THROW("zmq_bind failed on endpoint: %s, zmqerrno: %d",
+                m_endpoint.c_str(),
+                zmq_errno());
+    }
+
+    // zmq_poll will use less CPU
+    zmq_pollitem_t poll_item[1];
```
Comment (qiluo-msft) [inline common/zmqserver.cpp:115]: >poll_item [](http://example.com/codeflow?start=19&length=9)

Just one variable, not to use array?

File: common/zmqconsumerstatetable.cpp (hunk 18 above)
Comment (liuh-80) [reply common/zmqconsumerstatetable.cpp:80]: The default constructor not work, we need a deep copy here for the std::vector<FieldValueTuple> part:
typedef std::tuple<std::string, std::string, std::vector<FieldValueTuple> > KeyOpFieldsValuesTuple;

File: common/zmqserver.cpp (hunk 21 above)
Comment (liuh-80) [reply common/zmqserver.cpp:104]: ZmqServer side no need mutex because it does not access resource across thread, the ZmqConsumerStateTable will use mutex.

File: common/zmqserver.cpp (hunk 22 above)
Comment (liuh-80) [reply common/zmqserver.cpp:115]: Yes, the API design to input an array for poll multiple sockets. but we only have 1 socket:

http://api.zeromq.org/2-1:zmq-poll 
int zmq_poll (zmq_pollitem_t *items, int nitems, long timeout);

File: common/zmqconsumerstatetable.h (hunk 19 above)
Comment (liuh-80) [reply common/zmqconsumerstatetable.h]: Fixed

File: common/zmqserver.cpp [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+                                    const std::string dbName,
+                                    const std::string tableName,
+                                    ZmqMessageHandler* handler)
+{
+    auto dbResult = m_HandlerMap.insert(pair<string, map<string, ZmqMessageHandler*>>(dbName, map<string, ZmqMessageHandler*>()));
+    if (dbResult.second) {
+        SWSS_LOG_DEBUG("ZmqServer add handler mapping for db: %s", dbName.c_str());
+    }
+
+    auto tableResult = m_HandlerMap[dbName].insert(pair<string, ZmqMessageHandler*>(tableName, handler));
```
Comment (qiluo-msft) [inline common/zmqserver.cpp:39]: >m_HandlerMap[dbName] [](http://example.com/codeflow?start=23&length=20)

To clarify, the benefit of using `insert` is that you will get the Boolean and iterator at the same time, you do not need index again as `m_HandlerMap[dbName]`. It will optimize the speed a little bit.

File: common/zmqconsumerstatetable.cpp (hunk 18 above)
Comment (qiluo-msft) [reply common/zmqconsumerstatetable.cpp:80]: The default constructor should be a deep copy, for vector and tuple. Right?

File: common/zmqserver.cpp (hunk 22 above)
Comment (qiluo-msft) [reply common/zmqserver.cpp:112]: I understand. If you use single var `poll_item`, you may use `zmq_poll(&poll_item, 1, 1000)`. I think this simplifies the code.

File: common/zmqserver.cpp (hunk 23 above)
Comment (liuh-80) [reply common/zmqserver.cpp:39]: I try use the iterator but the handler seems not add to the mapping correfctly: the add code successed, but when search again seems can't found.

File: common/zmqserver.cpp (hunk 22 above)
Comment (liuh-80) [reply common/zmqserver.cpp]: Fixed.

File: common/zmqconsumerstatetable.cpp (hunk 18 above)
Comment (liuh-80) [reply common/zmqconsumerstatetable.cpp]: Fixed, confirmed the ctor can do the deep copy.

**Review by qiluo-msft — DISMISSED** [review]

**Comment by liuh-80** [comment]
swss-common PR been blocked by this change in sairedis: https://github.com/sonic-net/sonic-sairedis/pull/1230
That PR add new UT and require modify azure pipeline for UT, will sync with PR owner to fix swss-common repo.

**Comment by jiahua-wang** [comment]
https://github.com/sonic-net/sonic-swss-common/pull/778 should help passing the sairedis syncd unit test.

File: common/zmqclient.cpp [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+        {
+            m_connected = false;
+            auto message =  "zmq connection break, endpoint: " + m_endpoint + ",error: " + to_string(rc);
+            SWSS_LOG_ERROR("%s", message.c_str());
+            throw system_error(make_error_code(errc::connection_reset), message);
+        }
+        else
+        {
+            // for other error, send failed immediately.
+            auto message =  "zmq send failed, endpoint: " + m_endpoint + ",error: " + to_string(rc);
```
Comment (qiluo-msft) [inline common/zmqclient.cpp:166]: >error [](http://example.com/codeflow?start=75&length=5)

Do you want add a blank before `error`?

File: common/zmqclient.cpp [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+        }
+        else if (zmq_err == EAGAIN)
+        {
+            // EAGAIN: ZMQ is full to need try again
+            SWSS_LOG_WARN("zmq is full, will retry in %d ms, endpoint: %s, error: %d", retry_delay, m_endpoint.c_str(), zmq_err);
+        }
+        else if (zmq_err == ETERM)
+        {
+            m_connected = false;
+            auto message =  "zmq connection break, endpoint: " + m_endpoint + ",error: " + to_string(rc);
```
Comment (qiluo-msft) [inline common/zmqclient.cpp:159]: >error [](http://example.com/codeflow?start=80&length=5)

Do you want add a blank before `error`?

File: common/table.h [hunk 26]
```diff
@@ -39,29 +39,34 @@ class TableBase {
...
+
+    TableBase(const std::string &tableName, const std::string &tableSeparator)
+        : m_tableName(tableName), m_tableSeparator(tableSeparator)
+    {
+        static const std::string legalSeparators = ":|";
+        if (legalSeparators.find(tableSeparator) == std::string::npos)
+            throw std::invalid_argument("Invalid table name separator");
+    }
+
+    static std::string gettableSeparator(int dbId)
```
Comment (qiluo-msft) [inline common/table.h:54]: >gettableSeparator [](http://example.com/codeflow?start=23&length=17)

gettableSeparator -> getTableSeparator

File: common/zmqclient.cpp (hunk 24 above)
Comment (liuh-80) [reply common/zmqclient.cpp]: Fixed

File: common/zmqclient.cpp (hunk 25 above)
Comment (liuh-80) [reply common/zmqclient.cpp]: Fixed

File: common/table.h (hunk 26 above)
Comment (liuh-80) [reply common/table.h]: Fixed
Also change ZmqConsumerTable inherite TableBase

**Review by qiluo-msft — APPROVED** [review]
